DM-17496: query size and data ID expansion optimizations

TallJimbo opened the pull request:
This change reduces the number of table clauses in the big `selectDimensions` query used in preflight, getting it below SQLite's maximum number of joins (64) in our tests (and hopefully all realistic cases).

~The remaining commits attempt to speed up some of the data ID manipulations we apply to each result row of that query, which turn out to be much slower than the query itself.  In my profiling, that brought down time spent in preflight for ci_hsc from about 12 minutes to about 5 minutes.  That's still much too slow, but my remaining plans for speedups involve some combination of more significant refactoring and lifting code to C++, and I don't think either belongs on this ticket.~

andy-slac reviewed (approved):
Looks good. I'd probably keep `_buildDatasetSubquery` unchanged for now if there is a chance we could still use it for something else, and just pass `False` for its `isOutput` argument. There is a separate issue with output dataset_id, I'll comment on JIRA.

TallJimbo commented:
For anyone following along at home, this branch has at least temporarily become something of an integration branch for various bugfixes that have come up in preparing for the demo, not all of them related to the original ticket goals.  We'll clean that up when we have some breathing room (possibly after the demo).

TallJimbo commented:
Good news: the 5 minutes reported above seems to have included at *lot* of profiler overhead; I just timed preflight for a slightly larger graph (added forced photometry tasks, relative to the above), and it came in at 1m30s.

TallJimbo committed:
Transform selectDimensions args into SqlPreFlight attributes.

This gives all SqlPreFlight methods access to the arguments,
reducing the need to change method signatures when changing
the relative responsibilities of private methods.

TallJimbo committed:
Add option to defer output dataset_id queries.

This should let us fit within the SQLite join limit for essentially
and real-world selectDimensions query.

TallJimbo committed:
Pass data ID through Registry.find and getDataset.

This ensures metadata associated with the data ID is not lost.

andy-slac commented on python/lsst/daf/butler/registries/sqlPreFlight.py:
```diff
@@ -30,6 +30,12 @@
 
 
+class DATASET_ID_DEFERRED:  # noqa N801
+    """Type used as a sentinal value to indicates when a dataset ID was not
```
indicates -> indicate?

andy-slac commented on python/lsst/daf/butler/registries/sqlPreFlight.py:
```diff
@@ -44,12 +50,39 @@ class SqlPreFlight:
+        The `list` of `DatasetTypes <DatasetType>` whose Dimensions will
+        be included in the returned column set. It is expected that
+        Datasets for these DatasetTypes do not exist in the registry,
+        but presently this is not checked.
```
This second sentence probably does not make much sense any more, that check is done in `GraphBuilder`, this class is not supposed to do that check.

andy-slac commented on python/lsst/daf/butler/registries/sqlPreFlight.py:
```diff
@@ -279,15 +297,19 @@ def findSkyPixSubstitute():
                        "output" if isOutput else "input", dsType.name)
 
+            if isOutput and self.deferOutputIdQueries:
+                dsIdColumns[dsType] = DATASET_ID_DEFERRED
```
Maybe add `_LOG.debug` here to get some diagnostics about this shortcut

andy-slac commented on tests/test_sqlPreFlight.py:
```diff
@@ -126,33 +125,33 @@ def testPreFlightInstrumentUnits(self):
+        for deferOutputIdQueries in (True, False):
+            with self.subTest(deferOutputIdQueries=deferOutputIdQueries):
+                preFlight = SqlPreFlight(self.registry, neededDatasetTypes=[rawType],
+                                         futureDatasetTypes=[calexpType], originInfo=originInfo)
```
Do you want to pass `deferOutputIdQueries` as argument to `SqlPreFlight` too?

TallJimbo replied:
Yes, oops.  Looks like I wasn't testing what I thought I was testing, but happily it does still pass with it fixed.

andy-slac reviewed (approved):
Looks OK, few minor comments.
